MobileWallet/Screens/Home/Transaction/TransactionViewController.swift

```diff
@@ -40,34 +40,102 @@
 
 import UIKit
 
-class TransactionViewController: UIViewController {
-    @IBOutlet weak var valueContainerView: UIView!
-    @IBOutlet weak var valueLabel: UILabel!
-    @IBOutlet weak var currencySymbol: UIImageView!
-    @IBOutlet weak var detailsStackView: UIStackView!
-
-    @IBOutlet weak var fromLabel: UILabel!
-    @IBOutlet weak var fromUserNameLabel: UILabel!
-    @IBOutlet weak var fromUserIdLabel: UILabel!
-
-    @IBOutlet weak var noteLabel: UILabel!
-    @IBOutlet weak var noteValueLabel: UILabel!
-    @IBOutlet weak var transactionIcon: UIImageView!
+class TransactionViewController: UIViewController, UITextFieldDelegate {
+    let SIDE_PADDING: CGFloat = 25
+    let BOTTOM_HEADING_PADDING: CGFloat = 20
+    let VALUE_VIEW_HEIGHT_MULTIPLIER_FULL: CGFloat = 0.21
+    let VALUE_VIEW_HEIGHT_MULTIPLIER_SHORTEND: CGFloat = 0.18
+
+    var contactPublicKey: PublicKey?
+    var contactAlias: String = ""
+    let dateContainerView = UIView()
+    let dateLabel = UILabel()
+    let valueContainerView = UIView()
+    var valueContainerViewHeightConstraintFull = NSLayoutConstraint()
+    var valueContainerViewHeightConstraintShortened = NSLayoutConstraint()
+    var valueCenterYAnchorConstraint = NSLayoutConstraint()
+    let valueLabel = UILabel()
+    let emojiButton = EmojiButton()
+    let fromHeadingLabel = UILabel()
+    let addContactButton = TextButton()
+    let contactNameHeadingLabel = UILabel()
+    let contactNameTextField = UITextField()
+    let editContactNameButton = TextButton()
+    let dividerView = UIView()
+    let noteHeadingLabel = UILabel()
+    let noteLabel = UILabel()
+    var noteHeadingLabelTopAnchorConstraintContactNameShowing = NSLayoutConstraint()
+    var noteHeadingLabelTopAnchorConstraintContactNameMissing = NSLayoutConstraint()
+
+    @IBOutlet weak var transactionIDLabel: UILabel!
+
+    var isShowingContactAlias: Bool = true {
+        didSet {
+            if isShowingContactAlias {
+                noteHeadingLabelTopAnchorConstraintContactNameMissing.isActive = false
+                noteHeadingLabelTopAnchorConstraintContactNameShowing.isActive = true
+                addContactButton.isHidden = true
+                DispatchQueue.main.asyncAfter(deadline: .now() + 0.15, execute: {
+                    self.contactNameTextField.isHidden = false
```

Comment:
add [ weak self] in
guard let self = self else { return } to avoid return cycle

Reply:
Thanks, updated 👌